### PR TITLE
Fix throwing table key violation errors during hash collisions

### LINKED ISSUE #35046: Invalid duplicate key error for non duplicate keys
**Description:**
$title.

**Steps to reproduce:**
```ballerina
type Record record {
    readonly string x;
    readonly string y;
    int z;
};

table<Record> key(x, y) _ = table [
        {
            x: "x4",
            y: "2022/02/10",
            z: 100
        },
        {
            x: "x1",
            y: "2022/02/40",
            z: 100
        }
];
```

Both the `x` and `y` fields have different values in the two mapping constructor expressions, but this results in a duplicate key error.

```cmd
$ bal run x.bal 
Compiling source
        x.bal
ERROR [x.bal:(13:9,17:10)] duplicate key found in table row key('x, y') : 'x1, 2022/02/40'
```

**Affected Versions:**
2022.0.0

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #35046

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 1]
```diff
@@ -1132,6 +1315,10 @@ private Integer getTypeHash(BType type) {
         return Objects.hash(type.tag, type.name);
     }
 
+    private boolean getTypeEquality(BType typeA, BType typeB) {
+        return typeA.tag == typeB.tag && Objects.equals(typeA.name, typeB.name);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1319]: Will this logic work for subtypes and referedTypes?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 2]
```diff
@@ -996,19 +996,34 @@ public void visit(BLangTableConstructorExpr tableConstructorExpr) {
...
+                                        List<BLangExpression> keyArray, int hash) {
+        List<BLangExpression> existingExpList = keyValues.get(hash);
+        if (existingExpList.size() == keyArray.size()) {
+            boolean isEqual = true;
+            for (int i = 0; i < keyArray.size(); i++) {
+                isEqual = isEqual && equality(keyArray.get(i), existingExpList.get(i));
+            }
+            return isEqual;
+        }
+        return false;
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1024]:
```suggestion
    private boolean checkForKeyEquality(HashMap<Integer, List<BLangExpression>> keyValues,
                                        List<BLangExpression> keyArray, int hash) {
        List<BLangExpression> existingExpList = keyValues.get(hash);
        boolean isEqual = false;
        if (existingExpList.size() == keyArray.size()) {
            isEqual = true;
            for (int i = 0; i < keyArray.size(); i++) {
                isEqual = isEqual && equality(keyArray.get(i), existingExpList.get(i));
            }
        }
        return isEqual;
```

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1024]: This is just to workaround the code coverage

Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1024]:
```suggestion
        List<BLangExpression> existingExpList = keyValues.get(hash);
        boolean isEqual = false;
        if (existingExpList.size() == keyArray.size()) {
            isEqual = true;
            for (int i = 0; i < keyArray.size(); i++) {
                isEqual = isEqual && equality(keyArray.get(i), existingExpList.get(i));
            }
        }
        return isEqual;
```

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1024]: Seems 1024 is never hit by codecov,

Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/35130/commits/34c4d787ea2d75f7020af780491c470b7df32f25

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java (hunk 1 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java]: Updated to chek for assignability instead in https://github.com/ballerina-platform/ballerina-lang/pull/35130/commits/34c4d787ea2d75f7020af780491c470b7df32f25

**Review by gimantha — APPROVED** [review]
LGTM

**Pull request merged**